### ISSUE TITLE
unstable column order in rows.Scan() binding?

### ISSUE THREAD

**Issue by jchris** [couchbase/go_n1ql#3]
Apologies for the screenshots. I am wrapping up my experiments with this because I think the overhead in conforming to the sql.DB interface is unnecessary in my case. However I noticed odd behavior where the bindings in Scan don't seem to be stable. I can't really trace it down, maybe it's happening in the sql.DB standard library code?

You see the key and value being reversed across refreshes. I am calling `stmt, err := db.N1QLConnection.Prepare(query)` with each HTTP request to my server (not recommended but maybe part of reproducing what I see.) In the screenshots the JSON `key` should always correspond to the meta object, but it sometimes doesn't, and I can't figure out why.

I am calling `rows.Scan(&meta, &value)` to get the results of a query like `SELECT META(default), beer from default WHERE beer IS NOT NULL` but my results vary across browser refresh. It's like sometimes `meta` is getting bound to the first result column and sometimes they are switched. I haven't tried with results with more than two columns yet.

<img width="682" alt="screen shot 2015-09-25 at 6 42 21 am" src="https://cloud.githubusercontent.com/assets/253/10102092/7f82d29e-6351-11e5-9f81-3c6f6d0867d5.png">

<img width="676" alt="screen shot 2015-09-25 at 6 42 45 am" src="https://cloud.githubusercontent.com/assets/253/10102097/89a39cd6-6351-11e5-9d34-3a6718e5d6de.png">

Here is a screenshot of my code, and [a link to the branch.](https://github.com/couchbase/sync_gateway/tree/feature/n1ql)

<img width="561" alt="screen shot 2015-09-25 at 6 43 01 am" src="https://cloud.githubusercontent.com/assets/253/10102131/a0cbefee-6351-11e5-9858-66a87514b729.png">

I was able to reproduce the screenshots from fresh material as gists https://gist.github.com/jchris/a921a72f102aa5210d04

**Comment by jchris** [comment]
I tried putting an `as` statement into my query but the problem persists.

``` sql
SELECT META(default) as meta, beer from default WHERE beer IS NOT NULL
```

**Comment by geraldss** [comment]
Hi Chris,

We need to provide an additional API for use by our new shell. We should make sure that API meets your needs as well.

We will send you a proposal when ready.

Gerald

Sent from my T-Mobile 4G LTE Device

-------- Original message --------

**Comment by jchris** [comment]
What I need is pretty simple.

On config change, I'll validate and prepare some statements. At query time I will run them and filter the results to match the user's channels. Validation consists of making sure the query starts with selecting `meta.id` and probably current rev information from the `_sync` field, which I need for filtering.

It looks like this library has all the code that we need for service discovery etc, but breaking it free of the sql.DB interface would make it easier to adapt for our purposes, I wouldn't need to care how many Columns the query selects, etc.

**Comment by jchris** [comment]
Got something running with gocb. Thanks!

**Comment by maniktaneja** [comment]
@jchris  Thanks for reporting this issue. Fixed by d3910182e5c925127a1424246ff0cba8c1d80108
